clusterversion: clarify where to add new version gates during stability

celiala opened the pull request:
The previous direction said to always put append version gates to the end
of the list. However, during the stability period, before we mint the
release branch, there are two places to add new gates:

- new gates should go _before_ the "mint" version, if it is being backported
  to the release branch and we have not yet merged the "mint" version to
  the release branch.
- new gates for the upcoming development, should be appended to the end
  of the list.

This direction should be reverted/simplified (to direct to always append) once the commit to "mint" 23.1 has been backported and merged to release-23.1. Tracking issue for this is: https://cockroachlabs.atlassian.net/browse/REL-311

Release note: None
Epic: REL-283

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99982)
<!-- Reviewable:end -->

celiala commented:
Confirmed that the changes for `V23_2StopWritingPayloadAndProgressToSystemJobs` to `V23_1StopWritingPayloadAndProgressToSystemJobs` match its backport[1] to release-23.1.

[1] https://github.com/cockroachdb/cockroach/pull/99937/files

celiala commented:
@rail @jlinder @dt - this is RFAL! 

It would be good to merge+backport this before more version gates go into the wrong place 😅 (another one merged last week).

celiala committed:
clusterversion: clarify where to add new version gates during stability

The previous direction said to always put append version gates to the end
of the list. However, during the stability period, before we mint the
release branch, there are two places to add new gates:

- new gates should go _before_ the "mint" version, if it is being backported
  to the release branch and we have not yet merged the "mint" version to
  the release branch.
- new gates for the upcoming development, should be appended to the end
  of the list.

This direction should be reverted/simplified (to direct to always append)
once the commit to "mint" 23.1 has been backported and merged to release-23.1.
Tracking issue for this is: cockroachlabs.atlassian.net/browse/REL-311

Release note: None
Epic: REL-283

rail reviewed (approved):
> Reviewed 7 of 8 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @dt)



<!-- Sent from Reviewable.io -->

celiala commented:
TFTR!

bors r=rail

celiala commented:
this does not need to be backported to release-23.1: 
- on release-23.1: version gates ordering/values are correct 
- on master: this PR added directions and corrected the master branch to match
